src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,692 @@
+        let hover_area_height = &height / 2.0 + &SHAPE_HOVER_AREA_SIZE.px();
+        let hover_area        = Rect((&hover_area_width,&hover_area_height));
+        let hover_area        = hover_area.translate_y(-hover_area_height/2.0);
+
+        let shrink           = 1.px() - 1.px() * grow;
+        let radius           = node::NODE_SHAPE_RADIUS.px();
+        let port_area_size   = SHAPE_MAX_WIDTH.px() * grow;
+        let port_area_width  = &width  + (&port_area_size - &shrink) * 2.0;
+        let port_area_height = &height + (&port_area_size - &shrink) * 2.0;
+        let bottom_radius    = &radius + &port_area_size;
+        let port_area        = Rect((&port_area_width,&port_area_height));
+        let port_area        = port_area.corners_radius(&bottom_radius);
+        let port_area        = port_area - BottomHalfPlane();
+        let corner_radius    = &port_area_size / 2.0;
+        let corner_offset    = &port_area_width / 2.0 - &corner_radius;
+        let corner           = Circle(&corner_radius);
+        let left_corner      = corner.translate_x(-&corner_offset);
+        let right_corner     = corner.translate_x(&corner_offset);
+        let port_area        = port_area + left_corner + right_corner;
+        let port_area        = port_area.into();
+        let hover_area       = hover_area.into();
+        BaseShapeData{port_area,hover_area,radius,width}
+    }
+}
+
+/// Trait that allows us to abstract the API of the `multi_port_area::Shape` and the
+/// `single_port_area::Shape`. This is needed to avoid code duplication for functionality that can
+/// work with either shape.
+#[allow(missing_docs)]
+trait PortShape {
+    fn set_grow(&self, grow_value:f32);
+    fn set_opacity(&self, opacity:f32);
+}
+
+
+
+// ========================
+// === Multi Port Shape ===
+// ========================
+
+/// Implements the shape for a segment of the OutputPort with multiple output ports.
+pub mod multi_port_area {
+    use super::*;
+    use ensogl::display::shape::*;
+    use std::f32::consts::PI;
+
+    /// Compute the angle perpendicular to the shape border.
+    fn compute_border_perpendicular_angle
+    (full_shape_border_length:&Var<f32>, corner_segment_length:&Var<f32>, position:&Var<f32>)
+        -> Var<f32> {
+        // TODO implement proper abstraction for non-branching "if/then/else" or "case" in
+        // shaderland
+        // Here we use a trick to use a pseudo-boolean float that is either 0 or 1 to multiply a
+        // value that should be returned, iff it's case is true. That way we can add return values
+        // of different "branches" of which exactly one will be non-zero.
+
+        // Transform coordinate system to be centered in the shape.
+        // The `distance`s here describe the distance along the shape border, so not straight line
+        // x coordinate, but the length of the path along the shape.
+        let center_distance          = position - full_shape_border_length  / 2.0;
```

Comment:
Why you have to "Transform coordinate system to be centered in the shape."? All the functions that are drawing shapes should accept arguments with coords already centered. Please lets keep this assumption so our life will be much easier :)